spirv_glsl.cpp

```diff
@@ -11813,11 +11868,52 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
 
 	// ALU
 	case OpIsNan:
-		GLSL_UFOP(isnan);
+		if (!is_legacy())
+			GLSL_UFOP(isnan);
+		else if (expression_type(ops[2]).vecsize > 1)
+			// Check if the number doesn't equal itself
+			emit_binary_func_op(ops[0], ops[1], ops[2], ops[2], "notEqual");
+		else
+			emit_binary_op(ops[0], ops[1], ops[2], ops[2], "!=");
 		break;
 
 	case OpIsInf:
-		GLSL_UFOP(isinf);
+		if (!is_legacy())
+			GLSL_UFOP(isinf);
+		else
+		{
+			// inf * 2 == inf by IEEE 754 rules, note this also applies to 0.0
+			// This is more reliable than checking if product with zero is NaN
+			uint32_t result_type = ops[0];
+			uint32_t result_id = ops[1];
+			uint32_t operand = ops[2];
+
+			auto &type = get<SPIRType>(result_type);
+			std::string expr;
+			if (type.vecsize > 1) {
+				expr = type_to_glsl_constructor(type);
+				expr += '(';
+				for (uint32_t i = 0; i < type.vecsize; i++)
+				{
+					auto comp = to_extract_component_expression(operand, i);
+					expr += join(comp, " != 0.0 && 2.0 * ", comp, " == ", comp);
+
+					if (i + 1 < type.vecsize)
+						expr += ", ";
+				}
+				expr += ')';
+			}
+			else
+			{
+				// Need explicit flush since we only call to_expression once
+				flush_variable_declaration(operand);
```

Comment:
flush_variable_declaration does not do what you expect here. track_expression_read() is probably what you want instead. You can call that after to_enclosed_expression.